esm: add import.meta.dirname and import.meta.filename (#48740)
nodejs/node: doc/api/esm.md

The `import.meta` meta property is an `Object` that contains the following
properties.

### `import.meta.__dirname`

<!-- YAML
added: REPLACEME
-->

* {string} The directory name of the current module. This is the same as the
  [`path.dirname()`][] of the [`import.meta.__filename`][].

> > **Caveat** only local modules support this property. Network modules will

Reviewer: ```suggestion
> **Caveat** only local modules support this property. Modules not using the `file:` protocol will
```

> not provide it.

### `import.meta.__filename`

<!-- YAML
added: REPLACEME
-->

* {string} The file name of the current module. This is the current module
  file's absolute path with symlinks resolved. This is the same as the
  [`url.fileURLToPath()`][] of the [`import.meta.url`][].

> > **Caveat** only local modules support this property. Network modules will

Reviewer: ```suggestion
> **Caveat** only local modules support this property. Modules not using the `file:` protocol will
```

The `import.meta` meta property is an `Object` that contains the following
properties.

### `import.meta.dirname`

<!-- YAML
added: REPLACEME
-->

> 

Reviewer 2: ```suggestion

> Stability: 1.2 - Release candidate
```

* {string} The directory name of the current module. This is the same as the
  [`path.dirname()`][] of the [`import.meta.filename`][].

> **Caveat** only local modules support this property. Modules not using the
> `file:` protocol will not provide it.

### `import.meta.filename`

<!-- YAML
added: REPLACEME
-->

> 

Reviewer 2: > Stability: 1.2 - Release candidate

The `import.meta` meta property is an `Object` that contains the following
properties.

### `import.meta.dirname`

<!-- YAML
added: REPLACEME
-->

> Stability: 1.2 - Release candidate

* {string} The directory name of the current module. This is the same as the
  [`path.dirname()`][] of the [`import.meta.filename`][].

> > **Caveat** only file-based modules support this property.

Reviewer 3: Unless I'm missing something, this is straight up false, a virtual module would have this property as long as its URL is under the `file` scheme.

```suggestion
> **Caveat**: only present on `file:` modules.
```

The `import.meta` meta property is an `Object` that contains the following
properties.

### `import.meta.dirname`

<!-- YAML
added: REPLACEME
-->

> Stability: 1.2 - Release candidate

* {string} The directory name of the current module. This is the same as the
  [`path.dirname()`][] of the [`import.meta.filename`][].

> **Caveat** only local modules support this property. Modules not using the

> > `file:` protocol will not provide it.

Reviewer 4: `local modules` might be misleading. This will break even for modules loaded from `localhost`, which arguably is local, or from in-memory data URLs.

Author: Suggestions are welcome. Note, "this will break" is inaccurate. Loading a module from the network, be it remote or local, is not supported by this feature as it is nonsensical.

Reviewer 2: ```suggestion
> **Caveat** only file-based modules support this property.
```

> ### `import.meta.filename`

Reviewer 4: I still agree with others that `filename` is misleading, especially considering that other runtimes implement `import.meta.file` as the actual file name and `import.meta.path` as the full path. (Then again, this entire API is a questionable fit for ESM I suppose.)

I assume this has been discussed and dismissed elsewhere, I am only mentioning it in case it hasn't.

Reviewer 2: It was discussed in the spec proposal. I summarized the conclusions here: https://github.com/wintercg/import-meta-registry/blob/main/import-meta-path-helpers.md#alternatives-considered

added: REPLACEME
-->

> > Stability: 1.2 - Release candidate

Reviewer 2: ```suggestion
> Stability: 1.2 - Release candidate

```

added: REPLACEME
-->

> > Stability: 1.2 - Release candidate

Reviewer 2: ```suggestion
> Stability: 1.2 - Release candidate

```

If needed, a `require` function can be constructed within an ES module using
[`module.createRequire()`][].

#### No `__filename` or `__dirname`

These CommonJS variables are not available in ES modules.

`__filename` and `__dirname` use cases can be replicated via
[`import.meta.url`][].

> 

Reviewer 2: I don't think we can fully delete this section; we need to edit it to explain that the variables `__filename` and `__dirname` don't exist as pseudo-globals but do exist as `import.meta.filename` and `import.meta.dirname`.

Reviewer 3: This still needs to be addressed

[`data:` URLs]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs
[`export`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export
[`import()`]: #import-expressions
[`import.meta.__dirname`]: #importmeta__dirname
[`import.meta.__filename`]: #importmeta__filename
[`import.meta.resolve`]: #importmetaresolvespecifier-parent
[`import.meta.url`]: #importmetaurl
[`import`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import
[`module.createRequire()`]: module.md#modulecreaterequirefilename
[`module.register()`]: module.md#moduleregister
[`module.syncBuiltinESMExports()`]: module.md#modulesyncbuiltinesmexports
[`package.json`]: packages.md#nodejs-packagejson-field-definitions

> [`path.dirname()`]: path.md#dirname

Reviewer 5: ```suggestion
[`path.dirname()`]: path.md#pathdirnamepath
```

[`port.ref()`]: https://nodejs.org/dist/latest-v17.x/docs/api/worker_threads.html#portref
[`port.unref()`]: https://nodejs.org/dist/latest-v17.x/docs/api/worker_threads.html#portunref
[`process.dlopen`]: process.md#processdlopenmodule-filename-flags
[`string`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String

> [`url.fileURLToPath()`]: url.md#fileurltopath

Reviewer 5: ```suggestion
[`url.fileURLToPath()`]: url.md#urlfileurltopathurl
```
